File: app/src/main/java/com/ndhunju/dailyjournal/controller/backup/GoogleDriveRestApiActivity.java
package com.ndhunju.dailyjournal.controller.backup;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.MenuItem;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.api.client.extensions.android.http.AndroidHttp;
import com.google.api.client.extensions.android.json.AndroidJsonFactory;
import com.google.api.client.googleapis.extensions.android.gms.auth.GoogleAccountCredential;
import com.google.api.services.drive.Drive;
import com.ndhunju.dailyjournal.R;
import com.ndhunju.dailyjournal.controller.BaseActivity;
import com.ndhunju.dailyjournal.controller.service.DriveServiceHelper;
import com.ndhunju.dailyjournal.controller.service.GoogleSignInHelper;
import com.ndhunju.dailyjournal.util.UtilsView;

import androidx.appcompat.widget.Toolbar;
import androidx.core.util.Pair;

import static com.ndhunju.dailyjournal.controller.service.DriveServiceHelper.OPERATION_STATUS_FAIL;
import static com.ndhunju.dailyjournal.controller.service.DriveServiceHelper.OPERATION_STATUS_SUCCESS;

/**
 * The main {@link Activity} for the Drive REST API functionality.
 */
public class GoogleDriveRestApiActivity extends BaseActivity {

    // Constants
    private static final String TAG = GoogleDriveRestApiActivity.class.getSimpleName();
    private static final int REQUEST_CODE_SIGN_IN = 1;
    private static final int REQUEST_CODE_ERROR_RESOLUTION = 2;

    // Member Variables
    private final GoogleSignInHelper googleSignInHelper = GoogleSignInHelper.get();
    private DriveServiceHelper mDriveServiceHelper;

    // View Variables
    private ProgressDialog connectionPd;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // Setup Toolbar
        Toolbar toolbar = findViewById(R.id.toolbar);
        if (toolbar != null) {
            setSupportActionBar(toolbar);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }

        // Wire views
        connectionPd = new ProgressDialog(getActivity());
        connectionPd.setMessage(String.format(
                getString(R.string.msg_connecting),
                getString(R.string.str_google_drive)
        ));
        connectionPd.setCanceledOnTouchOutside(true);
        connectionPd.setIndeterminate(true);
        connectionPd.setCancelable(true);
        connectionPd.show();

        int googleServiceStatus = GoogleApiAvailability.getInstance()
                .isGooglePlayServicesAvailable(this);
        switch (googleServiceStatus) {
            case ConnectionResult.SERVICE_MISSING:
            case ConnectionResult.SERVICE_VERSION_UPDATE_REQUIRED:
            case ConnectionResult.API_UNAVAILABLE:
            case ConnectionResult.SERVICE_DISABLED:
                GoogleApiAvailability.getInstance().getErrorDialog(
                        this,
                        googleServiceStatus,
                        REQUEST_CODE_ERROR_RESOLUTION
                ).show();
                break;
            case ConnectionResult.SUCCESS:
                // Authenticate user if not already.
                GoogleSignInAccount signInAccount = GoogleSignIn.getLastSignedInAccount(getContext());
                if (signInAccount != null
                        && !signInAccount.isExpired()
                        && (signInAccount.getGrantedScopes()
                                         .containsAll(googleSignInHelper.requiredScopesAsSet()))
                        && (DriveServiceHelper.getLastOperationStatus(this)
                            != OPERATION_STATUS_FAIL)) {
                    onSignedInToGoogleAccount(signInAccount);
                } else {
                    requestSignIn();
                }
                break;
        }
    }

    protected void showProgress(boolean showProgress, String message) {
        if (showProgress) {
            connectionPd.setProgressStyle(ProgressDialog.STYLE_SPINNER);
            connectionPd.setMessage(message);
            connectionPd.show();
        } else if (!TextUtils.isEmpty(message)) {
            connectionPd.setProgressDrawable(null);
            connectionPd.setMessage(message);
        } else {
            connectionPd.dismiss();
        }
    }

    /**
     * Shows {@code message} to the user in a dialog. When user acknowledges the message, finishes
     * current activity and passes {@code success} to previous activity.
     */
    protected void showEndResultToUser(String message, boolean success) {
        setResult(success ? Activity.RESULT_OK : Activity.RESULT_CANCELED);
        if (getActivity() != null && !getActivity().isFinishing()) {
            UtilsView.alert(getActivity(), message, (dialog, which) -> {
                setResult(success ? Activity.RESULT_OK : Activity.RESULT_CANCELED);
                finish();
            });
        }
    }

    /**
     * Starts a sign-in activity using {@link #REQUEST_CODE_SIGN_IN}.
     */
    protected void requestSignIn() {
        Log.d(TAG, "Requesting sign-in");
        showProgress(true, getString(R.string.msg_requesting_sign_in));

        GoogleSignInClient client = GoogleSignInHelper.get().getGoogleSigInClient(this);

        // The result of the sign-in Intent is handled in onActivityResult.
        startActivityForResult(client.getSignInIntent(), REQUEST_CODE_SIGN_IN);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent resultData) {
        switch (requestCode) {
            case REQUEST_CODE_SIGN_IN:
                if (resultCode == Activity.RESULT_OK && resultData != null) {
                    handleSignInResult(resultData);
                } else {
                    DriveServiceHelper.setLastOperationStatus(this, OPERATION_STATUS_FAIL);
                    showEndResultToUser(
                            getString(R.string.msg_error_g_drive_user_not_signed_in),
                            false
                    );
                }
                break;
            case REQUEST_CODE_ERROR_RESOLUTION:
                if (resultCode == Activity.RESULT_OK) {
                    requestSignIn();
                } else {
                    DriveServiceHelper.setLastOperationStatus(this, OPERATION_STATUS_FAIL);
                    showEndResultToUser(
                            getString(R.string.msg_error_g_drive_user_not_signed_in),
                            false
                    );
                }
                break;
        }

        super.onActivityResult(requestCode, resultCode, resultData);
    }

    /**
     * Handles the {@code result} of a completed sign-in activity initiated from {@link
     * #requestSignIn()}.
     */
    private void handleSignInResult(Intent result) {
        GoogleSignIn.getSignedInAccountFromIntent(result)
                .addOnSuccessListener(googleSignInAccount -> {
                    DriveServiceHelper.setLastOperationStatus(getContext(), OPERATION_STATUS_SUCCESS);
                    onSignedInToGoogleAccount(googleSignInAccount);
                })
                .addOnFailureListener(exception -> {
                    Log.e(TAG, "Failed sign in.", exception);
                    DriveServiceHelper.setLastOperationStatus(this, OPERATION_STATUS_FAIL);
                    showEndResultToUser(
                            getString(R.string.msg_error_g_drive_user_not_signed_in),
                            false
                    );
                });
    }

    private void onSignedInToGoogleAccount(GoogleSignInAccount googleAccount) {
        Log.d(TAG, "Sign in successful");
        // Use the authenticated account to sign in to the Drive service.
        GoogleAccountCredential credential = GoogleAccountCredential.usingOAuth2(
                this,
                googleSignInHelper.requiredScopesAsStringList()
        );

        credential.setSelectedAccount(googleAccount.getAccount());
        Drive googleDriveService = new Drive.Builder(
                AndroidHttp.newCompatibleTransport(),
                new AndroidJsonFactory(),
                credential
        ).setApplicationName(getString(R.string.app_name)).build();

        onSignedInToGoogleDrive(googleDriveService);
    }

    protected void onSignedInToGoogleDrive(Drive googleDriveService) {
        showProgress(false, null);

        if (getIntent().getBooleanExtra(BUNDLE_SHOULD_FINISH_ON_SIGN_IN, false)) {
            setResult(RESULT_OK);
            finish();
            return;
        }

        // The DriveServiceHelper encapsulates all REST API and SAF functionality.
        // Its instantiation is required before handling any onClick actions.
        mDriveServiceHelper = new DriveServiceHelper(googleDriveService);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    public DriveServiceHelper getDriveServiceHelper() {
        return mDriveServiceHelper;
    }

    public Context getContext() {
        return this;
    }

    public Activity getActivity() {
        return this;
    }

}

File: app/src/main/java/com/ndhunju/dailyjournal/controller/backup/GoogleDriveRestApiActivity.kt
package com.ndhunju.dailyjournal.controller.backup

import android.accounts.Account
import android.app.Activity
import android.app.ProgressDialog
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.MenuItem
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.widget.Toolbar
import com.firebase.ui.auth.AuthUI
import com.firebase.ui.auth.FirebaseAuthUIActivityResultContract
import com.firebase.ui.auth.data.model.FirebaseAuthUIAuthenticationResult
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.common.Scopes
import com.google.api.services.drive.Drive
import com.google.api.services.drive.DriveScopes
import com.google.firebase.auth.FirebaseAuth
import com.ndhunju.dailyjournal.R
import com.ndhunju.dailyjournal.controller.BaseActivity
import com.ndhunju.dailyjournal.controller.service.DriveServiceHelper
import com.ndhunju.dailyjournal.controller.service.GoogleAuthHelper
import com.ndhunju.dailyjournal.util.UtilsView
import kotlinx.coroutines.launch

/**
 * The main {@link Activity} for the Drive REST API functionality.
 */
open class GoogleDriveRestApiActivity : BaseActivity() {

    // Constants
    companion object {
        private const val TAG = "GoogleDriveRestApiActivity"
        private const val REQUEST_CODE_ERROR_RESOLUTION = "REQUEST_CODE_ERROR_RESOLUTION"

        /** Pass true for this key to finish this activity upon successful sign in to google drive **/
        const val BUNDLE_SHOULD_FINISH_ON_SIGN_IN = "BUNDLE_SHOULD_FINISH_ON_SIGN_IN"
    }

    // Member Variables
    private val googleSignInHelper = GoogleAuthHelper(this)
    private lateinit var mDriveServiceHelper: DriveServiceHelper

    // View Variables
    private lateinit var connectionPd: ProgressDialog

    private val signInLauncher = registerForActivityResult(
        FirebaseAuthUIActivityResultContract()
    ) { result ->
        onSignInResult(result)
    }

    private val resolutionLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result: ActivityResult ->
        if (result.resultCode == RESULT_OK) {
            val newStatus = GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(this)
            if (newStatus == ConnectionResult.SUCCESS) {
                if (googleSignInHelper.getCredential()?.selectedAccount != null) {
                    onSignedInToGoogleAccount(googleSignInHelper.getCredential()?.selectedAccount)
                } else {
                    requestSignIn()
                }
            } else {
                DriveServiceHelper.setLastOperationStatus(this, DriveServiceHelper.OPERATION_STATUS_FAIL)
                showEndResultToUser(
                    getString(R.string.msg_error_g_drive_user_not_signed_in),
                    false
                )
            }
        } else {
            DriveServiceHelper.setLastOperationStatus(this, DriveServiceHelper.OPERATION_STATUS_FAIL)
            showEndResultToUser(
                getString(R.string.msg_error_g_drive_user_not_signed_in),
                false
            )
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Setup Toolbar
        val toolbar: Toolbar? = findViewById(R.id.toolbar)
        toolbar?.let {
            setSupportActionBar(it)
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
        }

        // Wire views
        // TODO: Use AlertDialog to show progress
        connectionPd = ProgressDialog(this).apply {
            setMessage(getString(R.string.msg_connecting, getString(R.string.str_google_drive)))
            setCanceledOnTouchOutside(true)
            isIndeterminate = true
            setCancelable(true)
            show()
        }

        val account = googleSignInHelper.getCredential()?.selectedAccount

        if (account != null) {
            onSignedInToGoogleAccount(account)
        } else {
            val apiAvailability = GoogleApiAvailability.getInstance()
            if (googleSignInHelper.errorCode != null && apiAvailability.isUserResolvableError(googleSignInHelper.errorCode)) {
                val intent = apiAvailability.getErrorResolutionIntent(
                    this,
                    googleSignInHelper.errorCode,
                    REQUEST_CODE_ERROR_RESOLUTION
                )
                intent?.let { resolutionLauncher.launch(it) }
            } else {
                requestSignIn()
            }
        }
    }

    // TODO: Use AlertDialog to show progress
    fun showProgress(showProgress: Boolean, message: String?) {
        if (showProgress) {
            connectionPd.setProgressStyle(ProgressDialog.STYLE_SPINNER)
            connectionPd.setMessage(message)
            connectionPd.show()
        } else if (!TextUtils.isEmpty(message)) {
            connectionPd.setProgressDrawable(null)
            connectionPd.setMessage(message)
        } else {
            connectionPd.dismiss()
        }
    }

    /**
     * Shows {@code message} to the user in a dialog. When user acknowledges the message, finishes
     * current activity and passes {@code success} to previous activity.
     */
    fun showEndResultToUser(message: String, success: Boolean) {
        setResult(if (success) RESULT_OK else RESULT_CANCELED)
        if (!isFinishing) {
            UtilsView.alert(this, message) { _, _ ->
                setResult(if (success) RESULT_OK else RESULT_CANCELED)
                finish()
            }
        }
    }

    /**
     * Starts a sign-in activity using {@link #REQUEST_CODE_SIGN_IN}.
     */
    fun requestSignIn() {
        Log.d(TAG, "Requesting sign-in")
        showProgress(true, getString(R.string.msg_requesting_sign_in))

        // Use FirebaseUI Auth to start the sign-in flow with Google provider
        val providers = arrayListOf(
            AuthUI.IdpConfig.EmailBuilder().build(),
            AuthUI.IdpConfig.GoogleBuilder()
                .setScopes(
                    listOf(
                        DriveScopes.DRIVE_FILE,
                        Scopes.EMAIL,
                        Scopes.PROFILE
                    )
                )
                .build()
        )

        val signInIntent = AuthUI.getInstance()
            .createSignInIntentBuilder()
            .setAvailableProviders(providers)
            .setLogo(R.mipmap.ic_app)
            .setTheme(R.style.AppTheme)
            .build()
        signInLauncher.launch(signInIntent)
    }

    /**
     * Handles the result of the FirebaseUI Auth sign-in flow.
     */
    private fun onSignInResult(result: FirebaseAuthUIAuthenticationResult) {
        if (result.resultCode == RESULT_OK) {
            // Successfully signed in
            val user = FirebaseAuth.getInstance().currentUser
            if (user != null) {
                Log.d(TAG, "Sign in successful: ${user.email}")
                DriveServiceHelper.setLastOperationStatus(
                    this@GoogleDriveRestApiActivity,
                    DriveServiceHelper.OPERATION_STATUS_SUCCESS
                )
                val account = user.email?.let { Account(it, "com.google") }
                onSignedInToGoogleAccount(account)
            } else {
                Log.e(TAG, "Sign in succeeded but user is null")
                DriveServiceHelper.setLastOperationStatus(
                    this@GoogleDriveRestApiActivity,
                    DriveServiceHelper.OPERATION_STATUS_FAIL
                )
                showEndResultToUser(
                    getString(R.string.msg_error_g_drive_user_not_signed_in),
                    false
                )
            }
        } else {
            // Sign in failed
            val response = result.idpResponse
            if (response == null) {
                Log.d(TAG, "User cancelled sign-in")
            } else {
                Log.e(TAG, "Sign in error", response.error)
            }
            DriveServiceHelper.setLastOperationStatus(
                this@GoogleDriveRestApiActivity,
                DriveServiceHelper.OPERATION_STATUS_FAIL
            )
            showEndResultToUser(
                getString(R.string.msg_error_g_drive_user_not_signed_in),
                false
            )
        }
    }

    private fun onSignedInToGoogleAccount(googleAccount: Account?) {
        val googleDriveService = googleSignInHelper.getDriveService()
        if (googleDriveService != null) {
            Log.d(TAG, "Sign in successful")
            onSignedInToGoogleDrive(googleDriveService)
        } else {

            Log.d(TAG, "Sign in failed")
            DriveServiceHelper.setLastOperationStatus(
                this,
                DriveServiceHelper.OPERATION_STATUS_FAIL
            )
            showEndResultToUser(
                getString(R.string.msg_error_g_drive_user_not_signed_in),
                false
            )
        }
    }

    open fun onSignedInToGoogleDrive(googleDriveService: Drive) {

        showProgress(false, null)

        if (intent.getBooleanExtra(BUNDLE_SHOULD_FINISH_ON_SIGN_IN, false)) {
            setResult(RESULT_OK)
            finish()
            return
        }

        // The DriveServiceHelper encapsulates all REST API and SAF functionality.
        // Its instantiation is required before handling any onClick actions.
        mDriveServiceHelper = DriveServiceHelper(googleDriveService)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                onBackPressedDispatcher.onBackPressed()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    fun getDriveServiceHelper(): DriveServiceHelper = mDriveServiceHelper

    fun getContext(): Context = this

    fun getActivity(): Activity = this
}
